File: script/drawing.js
define('drawing', ['entities', 'images', 'level', 'settings'], function(entities, images, level, settings) {

    var can = document.querySelector('canvas');
    var ctx = can.getContext('2d');

    var PAUSE_TEXT = 'HESITATION';

    // Handle canvas sizing.
    var width = can.width;
    var height = can.height;

    function onresize() {
        can.width = width = document.body.clientWidth;
        can.height = height = document.body.clientHeight;
    }
    onresize();
    window.addEventListener('resize', onresize);

    var entitiesImg = null;
    images.waitFor('entities').done(function(img) {
        entitiesImg = img;
    });

    var uiMelonSize = settings.tile_size;
    function drawUI() {
        ctx.drawImage(
            entitiesImg,
            0, 0,
            8, 8,
            10, ctx.canvas.height - 8 - uiMelonSize,
            uiMelonSize, uiMelonSize
        );

        ctx.font = (uiMelonSize | 0) + 'px VT323';
        ctx.fillStyle = 'black';
        ctx.fillText(
            'x' + entities.registry[0].melonCount,
            12 + uiMelonSize + 10,
            ctx.canvas.height - 23
        );
        ctx.fillStyle = 'white';
        ctx.fillText(
            'x' + entities.registry[0].melonCount,
            10 + uiMelonSize + 10,
            ctx.canvas.height - 25
        );
    }

    return {
        draw: function() {
            ctx.imageSmoothingEnabled = false;
            ctx.mozImageSmoothingEnabled = false;
            ctx.webkitImageSmoothingEnabled = false;
            level.getCurrent().draw(ctx, drawUI);
        },
        drawPaused: function() {
            var modifierWidth = ctx.measureText(PAUSE_TEXT);

            ctx.font = '60px VT323';
            ctx.fillStyle = 'black';
            ctx.fillText(PAUSE_TEXT, ctx.canvas.width / 2 - modifierWidth.width / 2 + 4, ctx.canvas.height / 2 - 30 + 4);
            ctx.fillStyle = 'white';
            ctx.fillText(PAUSE_TEXT, ctx.canvas.width / 2 - modifierWidth.width / 2, ctx.canvas.height / 2 - 30);
        }
    };
});

File: script/drawing.dart
import 'dart:html';

import 'entities.dart' as entities;
import 'images.dart' as images;
import 'levels.dart' as levels;
import 'settings.dart' as settings;


var _can;
var _ctx;

var _width;
var _height;

var entitiesDrawable;

const PAUSE_TEXT = 'HESITATION';


void init() {
    _can = document.querySelector('canvas');
    _ctx = _can.getContext('2d');

    _width = _can.width;
    _height = _can.height;

    onResize(null);
    window.onResize.listen(onResize);

    entitiesDrawable = images.get('entities');

}

void onResize(Event e) {
    _can.width = _width = document.body.clientWidth;
    _can.height = _height = document.body.clientHeight;
}

void drawUI() {
    entitiesDrawable.draw((img) {
        _ctx.drawImageScaledFromSource(
            img,
            0, 0,
            8, 8,
            10, _height - 8 - settings.tile_size,
            settings.tile_size, settings.tile_size
        );
    });

    _ctx.font = settings.tile_size.toString() + 'px VT323';
    _ctx.fillStyle = 'black';
    _ctx.fillText(
        'x' + entities.registry[0].melonCount.toString(),
        12 + settings.tile_size + 10,
        _ctx.canvas.height - 23
    );
    _ctx.fillStyle = 'white';
    _ctx.fillText(
        'x' + entities.registry[0].melonCount.toString(),
        10 + settings.tile_size + 10,
        _ctx.canvas.height - 25
    );
}

void drawPaused() {
    var modifierWidth = _ctx.measureText(PAUSE_TEXT);

    _ctx.font = '60px VT323';
    _ctx.fillStyle = 'black';
    _ctx.fillText(PAUSE_TEXT, _width / 2 - modifierWidth.width / 2 + 4, _height / 2 - 30 + 4);
    _ctx.fillStyle = 'white';
    _ctx.fillText(PAUSE_TEXT, _width / 2 - modifierWidth.width / 2, _height / 2 - 30);
}

void draw() {
    _ctx.imageSmoothingEnabled = false;
    levels.getCurrent().draw(_ctx, drawUI);
}
